network/src/protocols/wire/handshake/v1/bitvec.rs

```diff
@@ -0,0 +1,213 @@
+/// 1. Marking a position as set, and
+/// 2. Checking if a position is set.
+/// Internally, it stores a vector of u8's.
+/// * The first 8 positions of the bit vector are encoded in the first element of the vector, the
+///   next 8 are encoded in the second element, and so on.
+/// * Each bit of a u8 is set to 1 if the position is set and to 0 if it's not.
+/// * We only allow setting positions upto u8::MAX. As a result, the size of the inner vector is
+///   limited to 32 (= 256 / 8).
+/// * Once a bit has been set, it cannot be unset. As a result, the inner vector cannot shrink.
+/// * The positions can be set in any order.
+/// * A position can set more than once -- it remains set after the first time.
+///
+/// # Examples:
+/// ```ignore
+/// let mut bv = BitVec::default();
+/// bv.set(2);
+/// bv.set(5);
+/// assert!(bv.is_set(2));
+/// assert!(bv.is_set(5));
+/// assert_eq!(false, bv.is_set(0));
+/// ```
+#[derive(Clone, Default, Debug, PartialEq, Serialize)]
+pub struct BitVec {
+    inner: Vec<u8>,
+}
+
+impl BitVec {
+    // TODO(abhayb): Remove after migration to new wire format.
+    #[allow(dead_code)]
+    // Sets the bit at position @pos.
+    pub fn set(&mut self, pos: u8) {
+        // This is optimised to: let bucket = pos << 3;
+        let bucket: usize = pos as usize / BUCKET_SIZE;
+        if self.inner.len() <= bucket {
+            self.inner.resize(bucket + 1, 0);
+        }
+        // This is optimized to: let bucket_pos = pos | 0x07;
+        let bucket_pos = pos as usize - (bucket * BUCKET_SIZE);
+        self.inner[bucket] |= 0x01 << bucket_pos;
+    }
+
+    // TODO(abhayb): Remove after migration to new wire format.
+    #[allow(dead_code)]
+    // Checks if the bit at position @pos is set.
+    pub fn is_set(&self, pos: u8) -> bool {
+        // This is optimised to: let bucket = pos << 3;
+        let bucket: usize = pos as usize / BUCKET_SIZE;
+        if self.inner.len() <= bucket {
+            return false;
+        }
+        // This is optimized to: let bucket_pos = pos | 0x07;
+        let bucket_pos = pos as usize - (bucket * BUCKET_SIZE);
+        (self.inner[bucket] & (0x01 << bucket_pos)) != 0
+    }
+
+    // TODO(abhayb): Remove after migration to new wire format.
+    #[allow(dead_code)]
+    // Performs a bitwise AND of the current bit vector with the supplied bit vector.
+    pub fn and(&self, other: &Self) -> Self {
+        let len = std::cmp::min(self.inner.len(), other.inner.len());
+        let mut ret = BitVec {
+            inner: Vec::with_capacity(len),
+        };
+        for i in 0..len {
+            ret.inner.push(self.inner[i] & other.inner[i]);
+        }
+        ret
+    }
+}
+
+// We impl custom deseriazation to ensure that the length of inner vector does not exceed 32 (=
+// 256 / 8).
+impl<'de> Deserialize<'de> for BitVec {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct BitVecVisitor;
```

Comment:
You can potentially simplify this block like the following: https://github.com/libra/libra/blob/master/crypto/crypto/src/hash.rs#L289

Personally I find the pointed to style more readable and succinct -- but it only works if you are wrapping one type in another.

Reply:
great point. Since we are wrapping a Vec<u8> here, this pattern is useful. I'm also changing `Vec<u8>` to `Bytes`, since serialization and deserialization are more efficient for Bytes than sequences.